docs/1 - Glm Intro.py

```diff
@@ -0,0 +1,296 @@
+
+np.set_printoptions(suppress=True)
+torch.set_printoptions(sci_mode=False)
+# -
+
+# # Introduction
+#
+# `foundry.glm` is a package that seeks to learn the parameters of various parametric distributions. It uses gradient descent to minimize the negative log-likehood of the observed data, as a linear function (with activation) of some covariates.
+#
+# minimize the negative log-likehood <-->? maximimum likelihood estimator
+#
+# In this notebook, we demonstrate the functionality of the `foundry.glm`. We start off with a couple of basic examples with synthetic data, followed by a quick experiment with the online news dataset (https://archive.ics.uci.edu/ml/datasets/online+news+popularity).
+#
+# The synthetic data demonstrations we show are the following:
+# 1. Linear Regression
+# 2. Exponential Regression
+# 3. Logistic Regression
+# 4. Multiclass Classification
+#
+
+# # Synthetic Data Demonstrations
+
+# ## Linear Regression
+#
+# Linear regression follows the model that $y \sim N(\vec{x}\theta, \sigma^2)$, where $N$ is the gaussian distribution. 
+#
+# In the following, we show how to do that with the `Glm` class.
+
+# +
+# Define the model
+linear_glm = Glm("gaussian")
+# More options (will be explained later):
+# linear_glm = Glm("gaussian", penalty=1E-5)
+# linear_glm = Glm("gaussian", penalty=1E-5, col_mapping={"loc": ["column1", "column2", "column3"]} 
+
+# Define some data
+bias_true = 3.
+X, y_true, coefs_true = make_regression(
+    n_samples=10**5, # X.shape[0]
+    n_features=30, # X.shape[1]
+    n_informative=10, # 20 features don't predict the outcome
+    bias=bias_true, # E[y | X = 0]
+    effective_rank=15, # Increase correlation of features, as if there were approx 15 independent ones. 
+    noise=0.1, # defines variance
+    coef=True, # give the real model coefficients
+    random_state=97,
+)
+
+# fit the model
+linear_glm.fit(X, y_true)
+# More options:
+# linear_glm.fit(X, y_true, sample_weight=torch.rand(y_true.shape[0]))
+# linear_glm.fit(X, y_true, max_iter = 10**5)
+
+# evaluate the model
+print(f"Linear model R2: {r2_score(y_true, linear_glm.predict(X))}")
+print(f"Reducible error R2: {r2_score(y_true, X @ coefs_true + bias_true)}")
+
+# what is the actual reducible error?
+print(f"Linear model log-prob: {linear_glm.score(X, y_true)}")
+print(f"Linear model log-prob: {linear_glm.score(X, X @ coefs_true + bias_true)}")
+
+# -
+
+# ### Modelling heteroscedastic errors
+#
+# Above we can see how the Glm class can do linear regression. Ordinary least squares modelling typically assumes that the output variance is homoscedastic - i.e. that the variance $\sigma^2$ is constant, and that the error term $\epsilon = y - X\theta \sim N(0, \sigma^2)$. Sometimes that may not be the case, and we want to model the error as well. 
+#
+# We demonstrate this by adding some more noise to the outputs. This noise has zero mean, but variance that is a function of X. 
+
+# +
```

Comment:
based on reading through these examples, it occurred to me we'd do well to have our own data-simulating function. I've added one on 64d5e224d098c13193047ecb9f0b350073ea58af